### DISCUSSION TITLE
Setting custom textures

### DISCUSSION THREAD

**Discussion by chrisboyce** [emilk/egui#130] [Q&A] [answered]
I see in the docs that the `Image` widget takes a `TextureId`, but I can't figure out how to actually set the texture. In the back end, I see the `set_user_texture` function in the `Painter` impl for `egui_glium` , but I can't figure out how to get access to that struct from within the `updated` method. I'm using the `equi_template` as the starting point.

**Comment by emilk** [comment, marked as answer]
You use the `tex_allocator()` method in `epi::Frame`: https://docs.rs/epi/0.8.0/epi/struct.Frame.html#method.tex_allocator

``` rust
let tex_allocator = frame.tex_allocator().unwrap();
let texture_id = tex_allocator.alloc_srgba_premultiplied(size, pixels);
```

**Reply by lucasw** [reply]
If the color values of `pixels` are changed (but not the size), does `alloc_srgba_premultiplied` have to be called every update?  That's what I'm doing now and it works, but I'm wondering if that is doing a lot of unnecessary work in addition to updating the texture.

**Reply by emilk** [reply]
As long as you free the previous texture it should be fine.

**Reply by vojty** [reply]
This might sound like completely stupid question, but where should I free the textures?
I'm trying to render 400 squares (8x8 pixels), which are constantly changing

EDIT: I'm dumb, ignore me, I don't need 400 textures, I can do it with just one :)

**Reply by emilk** [reply]
Just in case: `TextureAllocator` also has ` free` method: https://docs.rs/epi/0.9.0/epi/trait.TextureAllocator.html#tymethod.free
